editor/src/components/canvas/controls/select-mode/select-mode-hooks.tsx

```diff
@@ -955,38 +955,47 @@ export function useClearKeyboardInteraction(editorStoreRef: {
 }
 
 type KeyboardEventListener = (e: KeyboardEvent) => void
+type UnloadEventListener = (e: BeforeUnloadEvent) => void
 
 export function useClearStaticReparentInteraction(editorStoreRef: {
   readonly current: EditorStorePatched
 }): () => void {
   const dispatch = useDispatch()
-  const mouseDownHandlerRef = React.useRef<EventListener>(NO_OP)
+  const mouseDownHandlerRef = React.useRef<UnloadEventListener>(NO_OP)
   const keyDownHandlerRef = React.useRef<KeyboardEventListener>(NO_OP)
 
   const removeEventListeners = () => {
     window.removeEventListener('mousedown', mouseDownHandlerRef.current)
+    window.removeEventListener('beforeunload', mouseDownHandlerRef.current)
     window.removeEventListener('keydown', keyDownHandlerRef.current)
   }
 
   React.useEffect(() => removeEventListeners)
 
   return React.useCallback(() => {
-    mouseDownHandlerRef.current = () => {
+    mouseDownHandlerRef.current = (e) => {
       removeEventListeners()
-
       if (
         editorStoreRef.current.editor.canvas.interactionSession?.interactionData.type ===
         'STATIC_REPARENT'
       ) {
         dispatch([CanvasActions.clearInteractionSession(true)], 'everyone')
+        e.returnValue = 'Unsaved changes'
+        return 'Unsaved changes'
```

Comment:
Return value in a mouse event handler function? Is there a reason for this?

Reply:
This is because the event handler is being reused for the `beforeunload` event, which is very cheeky. The string return value there instructs the browser to show a dialog with that text (https://developer.mozilla.org/en-US/docs/Web/API/BeforeUnloadEvent)

Reply:
I renamed the handler ref itself in https://github.com/concrete-utopia/utopia/pull/3798/commits/a4647e861bc28eeddd54640e51cea63347c4ca72, this is necessary because the same handler is subscribed to `beforeunload`, and the way to show the built-in `Changes that you made may not be saved` popup is to return a non-null or undefined value (https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event#compatibility_notes)